### DISCUSSION TITLE
MinIO STS with keycloak

### DISCUSSION THREAD

**Discussion by Guoyoyo** [minio/minio#11393] [Q&A] [answered]
Hello, I add an issue for the first time, sorry if it causes any inconvenience.

I refer to MinIO official document and try to integrate STS with Keycloak.

I have successfully integrated LDAP with Keycloak and configured the reference setting with MinIO too. [(keycloak.md)](https://github.com/minio/minio/blob/master/docs/sts/keycloak.md)

Because I do not have permission to set the LDAP's users attribues showed in thefollowing, there is empty when I successfully login MinIO GUI through KeyCloak.

<img src="https://user-images.githubusercontent.com/17918702/89866470-2c448000-dbe2-11ea-8c0f-85745c22ef01.png" width="700">
<img src="https://user-images.githubusercontent.com/17918702/89866179-ade7de00-dbe1-11ea-8aaa-4203333a6cc9.png" width="700">
So far, it’s reasonable. After all, I didn’t add a policy to the user.

## Current Behavior

Due to LDAP includes user email information,  I thought of adding a policy named user's email(e.g. nick@gmail.com) to `mc` in advance. After that I went to Keycloak to configure Mappers to  add a new mapping rule showed in the below. In this way, when I login to MinIO through KeyCloak by AD account, it will match user with the UserEmail policy. 

<img src="https://user-images.githubusercontent.com/17918702/89866714-9826e880-dbe2-11ea-9328-a283b18d910b.png" width="700">

In this case, LDAP user's name is Nick. If `mc` exists a policy named nick@gmail.com, when Log in with OpenID successful, user will be redirected to MinIO page and logged in automatically. Then, user would see now the buckets and objects they have access to.

<img src="https://user-images.githubusercontent.com/17918702/89867206-6bbf9c00-dbe3-11ea-86a7-c70e017ab7fe.png" width="700">

If `mc` not exist a UserEmail policy(nick@gmail.com) in advance, it will be redirected to a weird page. It does not show error message and no option to log out. 

<img src="https://user-images.githubusercontent.com/17918702/89868606-ab878300-dbe5-11ea-8ec4-6ae632db66c2.png" width="700">

## Expected Behavior

Q1: Without permission to add attributes for LDAP/AD 's users, is there a way to configure keycloak?
Q2: Can I avoid user redirect to that page even I do not create their UserEmail policy in advance ? 

## Error Log

```
API: AssumeRoleWithWebIdentity()
Time: 16:44:16 UTC 08/09/2020
DeploymentID: d60246e0-197a-4c2b-a5cc-c695b20a02d3
RequestID: 1629A744AF5F7C38
RemoteHost: <ip>
Host: <ip>:<port>
UserAgent: Go-http-client/1.1
Error: Invalid arguments specified
       4: github.com/minio/minio@/cmd/sts-errors.go:55:cmd.writeSTSErrorResponse()
       3: github.com/minio/minio@/cmd/sts-handlers.go:365:cmd.(*stsAPIHandlers).AssumeRoleWithJWT()
       2: github.com/minio/minio@/cmd/sts-handlers.go:409:cmd.(*stsAPIHandlers).AssumeRoleWithWebIdentity()
       1: net/http/server.go:2007:http.HandlerFunc.ServeHTTP()

API: WebLoginSTS()
Time: 16:44:16 UTC 08/09/2020
DeploymentID: d60246e0-197a-4c2b-a5cc-c695b20a02d3
RemoteHost: <ip>
Host: <ip>:<port>
UserAgent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36
Error: 400 Bad Request
      21: github.com/minio/minio@/cmd/web-handlers.go:2266:cmd.toWebAPIError()
      20: github.com/minio/minio@/cmd/web-handlers.go:2137:cmd.toJSONError()
      19: github.com/minio/minio@/cmd/web-handlers.go:2121:cmd.(*webAPIHandlers).LoginSTS()
      18: reflect/value.go:460:reflect.Value.call()
      17: reflect/value.go:321:reflect.Value.Call()
      16: github.com/gorilla/rpc@v1.2.0/v2/server.go:220:v2.(*Server).ServeHTTP()
      15: github.com/gorilla/mux@v1.7.0/mux.go:212:mux.(*Router).ServeHTTP()
      14: github.com/minio/minio@/cmd/generic-handlers.go:747:cmd.customHeaderHandler.ServeHTTP()
      13: github.com/minio/minio@/cmd/generic-handlers.go:762:cmd.securityHeaderHandler.ServeHTTP()
      12: github.com/minio/minio@/cmd/generic-handlers.go:623:cmd.bucketForwardingHandler.ServeHTTP()
      11: github.com/minio/minio@/cmd/generic-handlers.go:607:cmd.requestValidityHandler.ServeHTTP()
      10: github.com/minio/minio@/cmd/generic-handlers.go:543:cmd.httpStatsHandler.ServeHTTP()
       9: github.com/minio/minio@/cmd/generic-handlers.go:67:cmd.requestSizeLimitHandler.ServeHTTP()
       8: github.com/minio/minio@/cmd/generic-handlers.go:92:cmd.requestHeaderSizeLimitHandler.ServeHTTP()
       7: github.com/minio/minio@/cmd/crossdomain-xml-handler.go:51:cmd.crossDomainPolicy.ServeHTTP()
       6: github.com/minio/minio@/cmd/generic-handlers.go:246:cmd.redirectHandler.ServeHTTP()
       5: github.com/minio/minio@/cmd/generic-handlers.go:314:cmd.minioReservedBucketHandler.ServeHTTP()
       4: github.com/minio/minio@/cmd/generic-handlers.go:273:cmd.cacheControlHandler.ServeHTTP()
       3: github.com/minio/minio@/cmd/generic-handlers.go:384:cmd.timeValidityHandler.ServeHTTP()
       2: github.com/rs/cors@v1.6.0/cors.go:207:cors.(*Cors).Handler.func1()
       1: net/http/server.go:2007:http.HandlerFunc.ServeHTTP()
```

## Your Environment

* Version used (`minio --version`):minio version RELEASE.2020-04-23T00-58-49Z
* Server setup and configuration: I'm using minio with s3 gateway by using docker-compose  
* Operating System and version (`uname -a`): Linux

**Comment by harshavardhana** [comment, marked as answer]
> Q1: Without permission to add attributes for LDAP/AD 's users, is there a way to configure keycloak?

No

> Q2: Can I avoid user redirect to that page even I do not create their UserEmail policy in advance ?

No
